### PR TITLE
adding fixes so transducer can work again

### PR DESCRIPTION
**Pull request by bonham79**
Merges:  https://github.com/CUNY-CL/yoyodyne/pull/233  https://github.com/CUNY-CL/yoyodyne/pull/197 

Fixes: https://github.com/CUNY-CL/yoyodyne/issues/192 https://github.com/CUNY-CL/yoyodyne/issues/191

Dependent on: https://github.com/CUNY-CL/maxwell/pull/17

Summary: I fixed maxwell so TQDM isn't a property of the SED parameters anymore, this allows pickling of the expert module again and thus allows multigpu training across the transducer.

In progress I changed how the expert module initializes so that it just copies the index vocabulary from the dataloader that's passed to it.  So now you just need to pass an index to the action vocabulary and everything is managed in the backend. This allows free initialization of expert modules from checkpoints and thus skips epochs of em when resuming from a checkpointing. (I  just do the same thing we do with indexes in which you write the sed parameters to the experiment directory and load it when initializing the model.)

I also added new flags so that you can just skip em training for the expert. This is for an upcoming change in which the transducer is no longer dependent on having an oracle function. (I've found through training that the SED actually doesn't add that much to training.) It also allows the creation of `dummy` experts that just hold the action vocabulary. Added error checks to prevent unsafe use. Feel free to point out more.

I also moved adams changes from  https://github.com/CUNY-CL/yoyodyne/pull/233 into the trainer so that there's no weird attribute managing going on in the init anymore. ( Turns out that checkpointing pickles that kwargs dict, so simply adding the action vocabulary was creating too large an embedding space.)

I ran experiments over the Polish data and was able to write to predictions fairly easily. Only major issue is that we're wasting some parameters on creating target vocabulary embeddings that will never be used. But that's a low bar on the efficiency stack.

### CODE REVIEW

File: yoyodyne/models/expert.py [hunk 1]
```diff
@@ -17,12 +17,17 @@
 from torch.utils import data
 
 from .. import defaults
+from ..data import indexes
```
Comment (kylebgorman) [inline yoyodyne/models/expert.py:20]: Do this as `from .. import data` then use `data.Index`. That's why I did [this](https://github.com/CUNY-CL/yoyodyne/blob/9cea4d4d483c3b473b6d5a5cdc0abd1a05095419/yoyodyne/data/__init__.py#L8).

**Review by kylebgorman — COMMENTED** [review]
Looks good to me but some questions and style notes. Thanks Travis.

File: yoyodyne/models/expert.py [hunk 2]
```diff
@@ -32,19 +37,24 @@ class ActionVocabulary:
...
             actions.ConditionalDel(),
             actions.ConditionalCopy(),
         ]
         self.start_vocab_idx = len(self.i2w)
-        if i2w:
-            self.i2w.extend(i2w)
         self.w2i = {w: i for i, w in enumerate(self.i2w)}
-        self.target_characters = set()
-        self.encode_actions([unk_idx])  # Sets unknown character decoding.
+        # Use index from dataset to create action vocabulary.
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:50]: Don't we use imperative tense for our comments?

File: yoyodyne/models/expert.py [hunk 3]
```diff
@@ -240,7 +250,9 @@ class Expert(abc.ABC):
     oracle_factor: int
     roll_in: int
 
-    def __init__(self, actions, aligner, oracle_factor=defaults.ORACLE_FACTOR):
+    def __init__(
+        self, actions, aligner=None, oracle_factor=defaults.ORACLE_FACTOR
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:254]: Mmm, thought process was allow creation of dummy experts, but realize this is unnecessary in our current cases. Will revert

File: yoyodyne/models/expert.py [hunk 4]
```diff
@@ -408,82 +431,64 @@ def get_expert(
     train_data: data.Dataset,
     epochs: int = defaults.ORACLE_EM_EPOCHS,
     oracle_factor: int = defaults.ORACLE_FACTOR,
-    sed_params_path: str = None,
+    sed_params_path: str = "",
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:434]: My understanding is for cases of typing we want the null values to be the equivalent nulls for the type, no?

File: yoyodyne/train.py [hunk 5]
```diff
@@ -204,7 +204,11 @@ def get_model_from_argparse_args(
             datamodule.train_dataloader().dataset,
             epochs=args.oracle_em_epochs,
             oracle_factor=args.oracle_factor,
-            sed_params_path=args.sed_params,
+            sed_params_path=(
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/train.py:207]: Correct, I missed some tinkering to avoid this case.

File: yoyodyne/models/transducer.py [hunk 6]
```diff
@@ -557,11 +560,19 @@ def predict_step(self, batch: Tuple[torch.tensor], batch_idx: int) -> Dict:
 
     def convert_prediction(self, prediction: List[List[int]]) -> torch.Tensor:
         """Converts prediction values to tensor for evaluator compatibility."""
+        # FIXME: the two steps below may be partially redundant.
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/transducer.py:563]: That's from Adam's PR, no comment.

File: yoyodyne/models/expert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/models/expert.py:50]: Nope, I think I even tried to go through and get rid of any lingering ones.

File: yoyodyne/models/transducer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/models/transducer.py:563]: Ah, let's remove it then since we don't know what it means.

[the remark this answers is not in the document]
Comment (Adamits) [reply yoyodyne/models/transducer.py:563]: I think this is copied from #233 

iirc I meant that looping and stacking predictions, and calling `util.pad_tensor_after_eos` may do some redundant things that could be cleaned up at some point.

File: yoyodyne/models/expert.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/models/expert.py:434]: I'd say yes if this was C. It doesn't matter much in Python and indeed we use `None` like this elsewhere in library because that's how `argparse` works. Here's an example: if you're training from scratch the arg that's a checkpoint to conintue training from will be `None` for instance.

`""` is also used in some circles to mean "read from stdin if this is an input file or write to stdout if this is an output file" and that's not intended here.

File: yoyodyne/train.py [hunk 7]
```diff
@@ -273,8 +277,16 @@ def get_model_from_argparse_args(
         source_attention_heads=args.source_attention_heads,
         source_encoder_cls=source_encoder_cls,
         start_idx=datamodule.index.start_idx,
-        target_vocab_size=datamodule.index.target_vocab_size,
-        vocab_size=datamodule.index.vocab_size,
+        target_vocab_size=(
```
Comment (Adamits) [inline yoyodyne/train.py:280]: So now we orchestrate the vocab sizes once in the trainer, which already has a handle on the initialized expert? This is much nicer than w/e I was trying to do before.

It still somehow feels clunky I think, but that is an effect of the single embeddings matrix updates not easily aligning to the expert.

File: yoyodyne/models/expert.py [hunk 8]
```diff
@@ -408,82 +431,64 @@ def get_expert(
...
-            _generate_data_and_encode_vocabulary(train_data, actions),
-            epochs=epochs,
+        if epochs:
+            sed_aligner = sed.StochasticEditDistance.fit_from_data(
+                _generate_data(train_data),
+                epochs=epochs,
+            )
+            sed_aligner.params.write_params(sed_params_path)
+        else:
+            sed_params = sed.ParamDict.read_params(sed_params_path)
```
Comment (Adamits) [inline yoyodyne/models/expert.py:485]: Am I understanding correctly that in order to use an existing sed, the user would specify `0` or `None` for `oracle_em_epochs`? Avoiding rerunning em every time is a great feature, but I wonder if we can think through a more intuitive user interface for it, this feels a bit buried. Maybe you've already put some thought into this though so please let me know if you think this is already a good interface.

Comment (Adamits) [reply yoyodyne/models/expert.py:485]: Oh sorry I see the comment in the method header. I missed it before.

File: yoyodyne/models/expert.py [hunk 9]
```diff
@@ -408,82 +431,64 @@ def get_expert(
...
     Args:
         data (data.Dataset): dataset for generating expert vocabulary.
         epochs (int): number of EM epochs.
-        sched_factor (float): scaling factor to determine rate of
+        oracle_factor (float): scaling factor to determine rate of
             expert rollout sampling.
+        sed_params_path (str): path to read/write location of sed parameters.
+            If epochs > 0, this is a write path.
+            If epochs == 0, this is a read path.
+            If empty string then creates 'dummy' expert.
```
Comment (Adamits) [inline yoyodyne/models/expert.py:446]: Maybe add a comment here or in the README about what behavior a 'dummy' expert entails.

**Comment by Adamits** [comment]
Thanks for these changes + cleaning this stuff up!! 

> Summary: I fixed maxwell so TQDM isn't a property of the SED parameters anymore, this allows pickling of the expert module again and thus allows multigpu training across the transducer.

Awesome!

> In progress I changed how the expert module initializes so that it just copies the index vocabulary from the dataloader that's passed to it. So now you just need to pass an index to the action vocabulary and everything is managed in the backend. This allows free initialization of expert modules from checkpoints and thus skips epochs of em when resuming from a checkpointing. (I just do the same thing we do with indexes in which you write the sed parameters to the experiment directory and load it when initializing the model.)

1. "is managed in the backend" means in the expert.py code? 2. Are the expert modules deterministic given a dataset? For the index, this is the case and it will be created and written every time you run an experiment I think right? I believe we only load it when it is explicitly asked for in the args (e.g. for inference), though correct me if i'm wrong. So here, I guess if we default to epochs > 0,  then we just need to put something in the README that you can re-use the expert by setting epochs = 0 (and then maybe we raise an error if there is not expert where we expect it to be in this case.).

> I also moved adams changes from https://github.com/CUNY-CL/yoyodyne/pull/233 into the trainer so that there's no weird attribute managing going on in the init anymore. ( Turns out that checkpointing pickles that kwargs dict, so simply adding the action vocabulary was creating too large an embedding space.)

Nice, thanks. Yeah I was worried about that but in a big rush trying to make a fix when I made the PR and then promplty disappeared. It looks like you cleaned up some naming, etc as well, so is it safe to assume that you have valdated that those are the correct changes, or should I do more testing?

> I ran experiments over the Polish data and was able to write to predictions fairly easily. Only major issue is that we're wasting some parameters on creating target vocabulary embeddings that will never be used. But that's a low bar on the efficiency stack.

How was the accuracy w/ and w/out features?

File: yoyodyne/models/expert.py [hunk 10]
```diff
@@ -408,82 +431,64 @@ def get_expert(
...
 ) -> Expert:
     """Generates expert object for training transducer.
 
     Args:
         data (data.Dataset): dataset for generating expert vocabulary.
         epochs (int): number of EM epochs.
-        sched_factor (float): scaling factor to determine rate of
+        oracle_factor (float): scaling factor to determine rate of
             expert rollout sampling.
+        sed_params_path (str): path to read/write location of sed parameters.
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:443]: What would be the best way to preserve formatting then? I want to have something along the lines of a case statement for readibility

File: yoyodyne/models/expert.py [hunk 11]
```diff
@@ -408,82 +431,64 @@ def get_expert(
...
-        for item in data:
-            # Ignores last symbol since EOW.
-            target = item.target.tolist()[:-1]
-            actions.encode_actions(target)
-            if data.index.tie_embeddings:
-                source = item.source.tolist()[1:-1]
-                actions.encode_actions(source)
-
-    actions = ActionVocabulary(unk_idx=train_data.index.unk_idx)
+        assert data.has_target, """Passed dataset with no target to expert
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:468]: This is more of a sanity check thing to me. Makes more sense to write a single assert line than defining an error class, doing a coinditional check, then raising. Just an economy thing I guess.

File: yoyodyne/models/transducer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/transducer.py:554]: So should we leave the comment to clean up later or just remove now?

File: yoyodyne/models/expert.py [hunk 12]
```diff
@@ -32,19 +37,24 @@ class ActionVocabulary:
     start_vocab_idx: int
     target_characters: Set[Any]
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:38]: Since it's dependent on maxwell, there's still some uses where youc an pass any form of sequences. Though i realize the variable name is already creating an assumption.

[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:38]: Ah, realized why i did that. I don't want this module to be too dependent on maxwell classes, so just using any to make it more generalizable.

[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:38]: Technically the edit actions can accept any hashable symbol. So strings, ints, even tuples. So any is a better representation of its coverage given that maxwell is also symbol agnostic.

File: yoyodyne/models/expert.py [hunk 13]
```diff
@@ -32,19 +37,24 @@ class ActionVocabulary:
...
             actions.ConditionalCopy(),
         ]
         self.start_vocab_idx = len(self.i2w)
-        if i2w:
-            self.i2w.extend(i2w)
         self.w2i = {w: i for i, w in enumerate(self.i2w)}
-        self.target_characters = set()
-        self.encode_actions([unk_idx])  # Sets unknown character decoding.
+        # Use index from dataset to create action vocabulary.
+        self.encode_actions([index(t) for t in index.target_vocabulary])
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:51]: Yeah, it nees to track all potential edit actions for a given symbol.

File: yoyodyne/models/expert.py [hunk 14]
```diff
@@ -408,82 +431,64 @@ def get_expert(
...
+    def _generate_data(
+        data: data.Dataset,
     ) -> Iterator[Tuple[List[int], List[int]]]:
-        """Function to manage data encoding while aligning SED."
+        """Helper function to manage data encoding for SED."
 
-        SED training over the default data sampling is expensive.
-        Training is quicker if tensors are converted to lists.
-        For efficiency, we encode action vocabulary simultaneously.
+        We want just the encodings without BOS or EOS tokens. This
```
[the remark this answers is not in the document]
Comment (bonham79) [reply yoyodyne/models/expert.py:457]: We just want the raw symbols string. We don't want to worry about BOS and EOS being encoded by maxwell. 

What comment would make sense to you? (I'm too familiar for a good user friendly doc string.)

File: yoyodyne/train.py (hunk 7 above)
Comment (bonham79) [reply yoyodyne/train.py:278]: Yeah it's still a bit clunky but at least the clunkiness is occuring in the train script (which is localized clumsiness). Probably a good next issue is to break up the train script some more so it's less of a monolith.

File: yoyodyne/models/expert.py (hunk 8 above)
Comment (bonham79) [reply yoyodyne/models/expert.py:485]: I just added a bool to the train script that's triggered if a sed file already exists. I think this is a bit cleaner. Thoughts?

File: yoyodyne/models/expert.py (hunk 9 above)
Comment (bonham79) [reply yoyodyne/models/expert.py:446]: Removed dummy expert. Not important for this go around.

**Comment by bonham79** [comment]
>     1. "is managed in the backend" means in the expert.py code? 2. Are the expert modules deterministic given a dataset? For the index, this is the case and it will be created and written every time you run an experiment I think right? I believe we only load it when it is explicitly asked for in the args (e.g. for inference), though correct me if i'm wrong. So here, I guess if we default to epochs > 0,  then we just need to put something in the README that you can re-use the expert by setting epochs = 0 (and then maybe we raise an error if there is not expert where we expect it to be in this case.).
> 
Yep, expert.py manages. No dataset is passed to expert anymore. It just reads from the index file. 

Indexes are pickled between runs no? So it should be deterministic with new initialization. Regardless, the expert is pickled now so the weights remain same between runs.

> 
> Nice, thanks. Yeah I was worried about that but in a big rush trying to make a fix when I made the PR and then promplty disappeared. It looks like you cleaned up some naming, etc as well, so is it safe to assume that you have valdated that those are the correct changes, or should I do more testing?
> 

I've validated but always a fan of triple checking.

> 
> How was the accuracy w/ and w/out features?

I forget off hand but they were on par with my usual runs with polish.

File: yoyodyne/models/expert.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kylebgorman) [reply yoyodyne/models/expert.py:434]: You can't within the arglist, so move it out of the arglist. Something like this:

```python

def get_expert(
    train_data: data.Dataset,
    epochs: int = defaults.ORACLE_EM_EPOCHS,
    oracle_factor: int = defaults.ORACLE_FACTOR,
    sed_params_path: str = "",
) -> Expert:
    """Generates expert object for training transducer.

    * If epochs > 0, sed_params_path is a write path.
    * If epochs == 0, sed_params_path is a read path.
    * If it is an empty string, then it creates a 'dummy' expert.

    Args:
        data (data.Dataset): dataset for generating expert vocabulary.
        epochs (int): number of EM epochs.
        oracle_factor (float): scaling factor to determine rate of
            expert rollout sampling.
        sed_params_path (str): path to read/write location of sed parameters.

    Returns:
        expert.Expert.
    """

```

**Comment by bonham79** [comment]
@kylebgorman @Adamits are there responses I missed at all? else i can work on linting for push

**Comment by kylebgorman** [comment]
> @kylebgorman @Adamits are there responses I missed at all? else i can work on linting for push

it’s LGTM modulo style/lint.

note that it’s ‘Any’ not ‘any’

**Comment by bonham79** [comment]
> > @kylebgorman @Adamits are there responses I missed at all? else i can work on linting for push
> 
> it’s LGTM modulo style/lint.
> 
> note that it’s ‘Any’ not ‘any’

Kk, will run linting later in afternoon for merge.

**Comment by bonham79** [comment]
@Adamits any pertinent issues limiting merge?

**Comment by Adamits** [comment]
> @Adamits any pertinent issues limiting merge?

Sorry, will go back through right now.

File: yoyodyne/models/expert.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Adamits) [reply yoyodyne/models/expert.py]: Ah, in that case target_characters is a strange name :) - but lets keep it for now.

File: yoyodyne/models/expert.py (hunk 8 above)
Comment (Adamits) [reply yoyodyne/models/expert.py]: Yeah that works. I am having trouble wrapping my head back around all of this now, so I think if we want further updates to this interface those can go in a later issue after we use the transducer more (which I need to do for a yoyodyne resubmission anyway).

**Review by Adamits — APPROVED** [review]
LGTM

**Comment by bonham79** [comment]
@kylebgorman checks passed. ready to merge. let's kill this issue

**Review by kylebgorman — APPROVED** [review]
LGTM

**Comment by kylebgorman** [comment]
Sorry for slow responses, was traveling without laptop most of the last week.
